File: Data Structures & Algorithms/insert-new-interval/submission-0.py
class Solution:
    def insert(self, intervals: List[List[int]], newInterval: List[int]) -> List[List[int]]:
        res = []
	
	
        for i in range(len(intervals)):
            # if new interval comes before all other intervals
            if newInterval[1] < intervals[i][0]:
                res.append(newInterval)
                return res + intervals[i:]
            # new interval slots in after the current interval
            elif newInterval[0] > intervals[i][1]:
                # so then we append the interval and move onto the next 
                res.append(intervals[i])
            # this means there is an overlap
            # the lower bound of new interval is <= intervals[i][1] 
            else:
                # merge
                newInterval[0] = min(newInterval[0], intervals[i][0])
                newInterval[1] = max(newInterval[1], intervals[i][1])
        
        # here it means that after all the merging, newInterval is 
        # the largest non overlapping interval
        res.append(newInterval)
        return res

File: Data Structures & Algorithms/non-overlapping-intervals/submission-0.cpp
class Solution 
{
public:
    int eraseOverlapIntervals(vector<vector<int>>& intervals) 
    {
        int len = static_cast<int>(intervals.size());
        std::sort(intervals.begin(), intervals.end());
        int end_bound = intervals[0][1];
        int res = 0;

        for (int i = 1; i < len; i++)
        {
            int start = intervals[i][0];
            int end = intervals[i][1];

            // if the current interval comes after our end_bound
            if (start >= end_bound)
            {
                end_bound = end;
            }
            else
            {
                res++;
                end_bound = std::min(end_bound, end);
            }
        }

        return res;
        
    }
};
